src/codecs/encoder.rs

```diff
@@ -4,19 +4,22 @@ use crate::{
 };
 use bytes::BytesMut;
 use codecs::{
-    encoding::{Error, Framer, FramingConfig, Serializer, SerializerConfig},
+    encoding::{Error, Framer, Serializer},
     NewlineDelimitedEncoder, RawMessageSerializer,
 };
-use serde::{Deserialize, Serialize};
+use tokio_util::codec::Encoder as _;
 
 #[derive(Debug, Clone)]
 /// An encoder that can encode structured events into byte frames.
-pub struct Encoder {
+pub struct Encoder<Framer>
+where
+    Framer: Clone,
```

Comment:
Why generics instead of just making framer an `Option`?

Reply:
Because this allows us to specialize the implementations for `impl tokio_util::codec::Encoder<Event> for Encoder<Framer>` and `impl tokio_util::codec::Encoder<Event> for Encoder<()>` at compile time such that they are branch free:

```rust
impl tokio_util::codec::Encoder<Event> for Encoder<Framer> {
    type Error = Error;

    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
        let len = buffer.len();
        let mut payload = buffer.split_off(len);

        self.serialize(event, &mut payload)?;

        // Frame the serialized event.
        self.framer.encode((), &mut payload).map_err(|error| {
            emit!(EncoderFramingFailed { error: &error });
            Error::FramingError(error)
        })?;

        buffer.unsplit(payload);

        Ok(())
    }
}
```

```rust
impl tokio_util::codec::Encoder<Event> for Encoder<()> {
    type Error = Error;

    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
        let len = buffer.len();
        let mut payload = buffer.split_off(len);

        self.serialize(event, &mut payload)?;

        buffer.unsplit(payload);

        Ok(())
    }
}
```

Meaning that when we know that no framer is used we don't need to check for its known absence, same (but inverse) when using an encoder with a known framer. Since this is a very hot path, the difference can be within the order of 5-10% throughput, as can also be seen in the benchmarks in https://github.com/vectordotdev/vector/pull/10684#issuecomment-1090337942.